### PR TITLE
[WIP] Add NodeOperatorRegistry limits features | DONT MERGE

### PR DESCRIPTION
**Pull request by loga4**
1. Feature1: added target limit, which set by DAO
2. Feature2: added stuck and forgiven keys to count penalized/unpenalized NodeOperator

## Feature1: added target limit, which set by DAO
Added `bool targetValidatorsLimitActive`, `uint64 targetValidatorsKeysCount`  `uint64 unavaliableKeysCount`

In the same time we have 2 different limits - `vettedKeysCount`(stakingLimit) and `targetValidatorsKeysCount`.  
If `targetValidatorsLimitActive` is disabled we use **vettedKeysCount** only.
If `targetValidatorsLimitActive` is enabled we calculate how many keys are unavailable:
- targetValidatorsKeysCount >= vettedKeysCount: 0 unavailable keys - It's mean that NodeOperator can use all available keys
- usedKeys < targetValidatorsKeysCount < vettedKeysCount: (vettedKeysCount - targetValidatorsKeysCount) unavailable keys
- targetValidatorsKeysCount <= usedKeys: (vettedKeysCount - usedKeys) unavailable keys

We need `unavailable keys` to calculate common available keys for operator/module.

AvailableKeys before:
```
readyToDepositValidatorsKeysCount = vettedSigningKeysCount - depositedSigningKeysCount
```

AvailableKeys After:
```
readyToDepositValidatorsKeysCount = vettedSigningKeysCount - depositedSigningKeysCount - unavailableKeys
```

## Feature2: added stuck and forgiven keys to count penalized/unpenalized NodeOperator
We add two counters for each NodeOperator: **stuckKeysCount**, **forgivenKeysCount**
**StuckKeysCount** - used to count the number of keys that we request to exit, but they expired the allocated time for this

ForgivenKeysCount - only DAO can be used. It is needed in case of adjustments. For example, if the NodeOperator cannot exit the key, he can directly send ETH for this and we update this counter.

### CODE REVIEW

**Comment by TheDZhon** [comment]
Please consider to add the following rule also:
```
if (stucked - refund) > 0 than ready to deposit := 0 for this NO
```

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 1]
```diff
@@ -957,15 +1139,22 @@ contract NodeOperatorsRegistry is INodeOperatorsRegistry, AragonApp, IStakingMod
...
         distributed = 0;
+        uint256 sharesAmount;
         for (uint256 idx = 0; idx < recipients.length; ++idx) {
-            stETH.transferShares(recipients[idx], shares[idx]);
-            distributed = distributed.add(shares[idx]);
-            emit RewardsDistributed(idx, shares[idx]);
+            sharesAmount = shares[idx];
+            if (sharesAmount > 0 && penalized[idx]) {
+                uint256 penalizedAmound = sharesAmount.div(2);
+                sharesAmount = sharesAmount.sub(penalizedAmound);
```
[the remark this answers is not in the document]
Comment (TheDZhon) [reply contracts/0.4.24/nos/NodeOperatorsRegistry.sol:1152]: should be called through the `SelfOwnedStETHBurner` I suppose
will be delivered with #526

File: contracts/0.4.24/test_helpers/NodeOperatorsRegistryMock.sol [hunk 2]
```diff
@@ -107,6 +107,24 @@ contract NodeOperatorsRegistryMock is NodeOperatorsRegistry {
         emit NodeOperatorAdded(id, _name, _rewardAddress, 0);
     }
 
+    function testing_setNodeOperatorLimits(
+        uint256 _nodeOperatorId, 
+        bool targetValidatorsLimitActive,
+        uint64 targetValidatorsKeysCount,
+        uint64 unavaliableKeysCount,
+        uint64 stuckSigningKeysCount,
+        uint64 forgivenSigningKeysCount
```
Comment (TheDZhon) [inline contracts/0.4.24/test_helpers/NodeOperatorsRegistryMock.sol:116]: maybe `compensated` or `refunded`, `reimbursed`, `renumerated`

File: contracts/0.4.24/interfaces/INodeOperatorsRegistry.sol [hunk 3]
```diff
@@ -56,6 +56,17 @@ interface INodeOperatorsRegistry {
             uint64 usedSigningKeys
         );
 
+    function getNodeOperatorLimits(uint256 _nodeOperatorId)
+        external
+        view
+        returns (
+            bool targetValidatorsLimitActive,
+            uint64 targetValidatorsKeysCount,
+            uint64 unavaliableKeysCount,
```
Comment (TheDZhon) [inline contracts/0.4.24/interfaces/INodeOperatorsRegistry.sol:65]: can't get the meaning of 'unavailable'
maybe 'outstanding'?

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 4]
```diff
@@ -78,6 +81,15 @@ contract NodeOperatorsRegistry is INodeOperatorsRegistry, AragonApp, IStakingMod
 
     bytes32 internal constant TOTAL_SIGNING_KEYS_STATS = keccak256("lido.NodeOperatorsRegistry.totalSigningKeysStats");
 
+    /// @dev DAO target limit, used to check how many keys shoud be go to exit
```
Comment (TheDZhon) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:84]: 'should be exited first to lower the node operator share in protocol'

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 5]
```diff
@@ -78,6 +81,15 @@ contract NodeOperatorsRegistry is INodeOperatorsRegistry, AragonApp, IStakingMod
...
+    ///      UINT64_MAX - unlim
+    ///      0 - all deposited keys 
+    ///      N < deposited keys -
+    ///      deposited < N < vetted - use (N-deposited) as available
```
Comment (TheDZhon) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:88]: sorry, maybe the comment is not clear enough

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 6]
```diff
@@ -375,6 +426,97 @@ contract NodeOperatorsRegistry is INodeOperatorsRegistry, AragonApp, IStakingMod
         emit ExitedSigningKeysCountChanged(_nodeOperatorId, _exitedValidatorsKeysCount);
     }
 
+    /// @notice Updates the limit of the validators that can be used for deposit by DAO
+    /// @param _nodeOperatorId Id of the node operator
+    /// @param _targetValidatorsKeysCount New number of EXITED validators of the node operator
```
Comment (TheDZhon) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:431]: is it really 'EXITED'?

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 7]
```diff
@@ -375,6 +426,97 @@ contract NodeOperatorsRegistry is INodeOperatorsRegistry, AragonApp, IStakingMod
...
+
+        _setTotalUnavailableKeysCount(totalUnavailableKeysCount);
+
+        emit TargetValidatorsLimitChanged(
+            _nodeOperatorId, 
+            _targetValidatorsKeysCount, 
+            _active, 
+            _nodeOperatorsLimits[_nodeOperatorId].unavaliableKeysCount
+        );
+        // emit TotalUnavaliableKeysCountChanged(totalUnavailableKeysCount);
```
Comment (TheDZhon) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:477]: dead code

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 8]
```diff
@@ -116,13 +128,29 @@ contract NodeOperatorsRegistry is INodeOperatorsRegistry, AragonApp, IStakingMod
...
+    /// @dev Node Operator parameters and internal state
+    struct NodeOperatorLimit {
+         /// @dev Flag indicating if the operator has dao limits
+        bool targetValidatorsLimitActive;
+        /// @dev DAO target limit for operator
+        uint64 targetValidatorsKeysCount;
+        /// @dev unavaliable keys count
+        uint64 unavaliableKeysCount;
+        /// @dev stuck keys count from oracle report
+        uint64 stuckSigningKeysCount;
```
Comment (TheDZhon) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:140]: why 'signing'?

**Review by TheDZhon — COMMENTED** [review]
Great to have this in the protocol :heart: 
Left some suggestions to improve readability

**Comment by TheDZhon** [comment]
please pick up trim for the single node operator

File: contracts/0.4.24/interfaces/INodeOperatorsRegistry.sol [hunk 9]
```diff
@@ -56,6 +56,17 @@ interface INodeOperatorsRegistry {
             uint64 usedSigningKeys
         );
 
+    function getNodeOperatorLimits(uint256 _nodeOperatorId)
```
Comment (avsetsin) [inline contracts/0.4.24/interfaces/INodeOperatorsRegistry.sol:59]: I propose to move this to the Staking Module interface. I believe these limits should apply to all modules

File: contracts/0.4.24/nos/NodeOperatorsRegistry.sol [hunk 10]
```diff
@@ -573,14 +715,46 @@ contract NodeOperatorsRegistry is INodeOperatorsRegistry, AragonApp, IStakingMod
         usedSigningKeys = nodeOperator.depositedSigningKeysCount;
     }
 
+    function getNodeOperatorLimits(uint256 _nodeOperatorId)
```
Comment (avsetsin) [inline contracts/0.4.24/nos/NodeOperatorsRegistry.sol:718]: Can we have an additional method that will return all limits for all node operators or in a bunch of `from` - `to`?

**Pull request closed without merging**